src/lib/client/jubSignalClient.ts

```diff
@@ -0,0 +1,296 @@
+  const keys = getKeys();
+  if (!profile || !keys) {
+    console.error("Error loading profile and keys from local storage");
+    throw new Error("Error loading profile and keys from local storage");
+  }
+
+  // Fetch jubSignal messages from server
+  const previousMessageFetchTime = session.lastMessageFetchTimestamp;
+  const newMessageFetchTime = new Date();
+  const urlStartFilter =
+    previousMessageFetchTime && !forceRefresh
+      ? `&startDate=${encodeURIComponent(
+          previousMessageFetchTime.toISOString()
+        )}`
+      : "";
+  const url =
+    `/api/messages?token=${encodeURIComponent(
+      session.authToken.value
+    )}&endDate=${encodeURIComponent(newMessageFetchTime.toISOString())}` +
+    urlStartFilter;
+  const response = await fetch(url, {
+    method: "GET",
+    headers: {
+      "Content-Type": "application/json",
+    },
+  });
+
+  if (!response.ok) {
+    console.error("Error fetching jubSignal messages from server");
+    throw new Error("Error fetching jubSignal messages from server");
+  }
+
+  // Decrypt messages and update localStorage with decrypted messages
+  // Start with empty users, location signatures, activities if forceRefresh is true
+  const messages = await response.json();
+  const existingUsers = forceRefresh ? {} : getUsers();
+  const existingLocationSignatures = forceRefresh
+    ? {}
+    : getLocationSignatures();
+  const existingActivities = forceRefresh ? [] : getActivities();
+  const { newUsers, newLocationSignatures, newActivities } =
+    await processEncryptedMessages({
+      messages,
+      recipientPrivateKey: keys.encryptionPrivateKey,
+      recipientPublicKey: profile.encryptionPublicKey,
+      existingUsers,
+      existingLocationSignatures,
+      existingActivities,
+    });
+
+  // Save users, location signatures, activities to localStorage
+  saveUsers(newUsers);
+  saveLocationSignatures(newLocationSignatures);
+  saveActivities(newActivities);
+
+  // Update the session
+  session.lastMessageFetchTimestamp = newMessageFetchTime;
+  saveSession(session);
+};
+
+// Helper function to process encrypted messages and update users, location signatures, activities
+const processEncryptedMessages = async (args: {
+  messages: any[];
+  recipientPrivateKey: string;
+  recipientPublicKey: string;
+  existingUsers: Record<string, User>;
+  existingLocationSignatures: Record<string, LocationSignature>;
+  existingActivities: Activity[];
+}): Promise<{
+  newUsers: Record<string, User>;
+  newLocationSignatures: Record<string, LocationSignature>;
+  newActivities: Activity[];
+}> => {
+  const {
+    messages,
+    recipientPrivateKey,
+    recipientPublicKey,
+    existingUsers: users,
+    existingLocationSignatures: locationSignatures,
+    existingActivities: activities,
+  } = args;
+
+  activities.reverse(); // We will reverse the activities array at the end - this is for faster array operations
+
+  for (const message of messages) {
```

Comment:
Are these guaranteed to be in reverse chronological order? Should we do some sort of sort at the end to verify? Sorting is cheap as hell

Reply:
I know the API endpoint sorts by created time but i'm more worried about multiple requests being fired off at the same time or some discrepancy between session saved request time and what is sent to server